### PR TITLE
FEATURE: Dashboard Page and package added framer motion

### PR DESCRIPTION
**Pull request by sahsisunny**
**Pull Request Description:**

### Changes Made:

1. **URL Dashboard Implementation:**
   - Introduced a new URL dashboard feature that displays a list of URLs created by users.
   - Each entry in the dashboard includes details such as the original URL, short URL, and the date it was created.

2. **Toast Functionality Fix:**
   - Addressed an existing issue with the toast functionality.
   - The fix ensures that toasts are displayed correctly and provide the expected user feedback.

### Testing:
File                        | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s
----------------------------|---------|----------|---------|---------|-------------------------------
All files                   |   86.17 |    72.38 |   71.66 |   86.01 |                               
 src/components/Dashboard/UrlListItem.tsx           |     100 |      100 |     100 |     100 | 
 src/components/Navbar/index.tsx                 |   94.44 |    66.66 |      75 |   94.11 | 31
 src/components/Toast/index.tsx                 |     100 |      100 |     100 |     100 | 
 src/pages/app/index.tsx                 |   56.52 |    36.36 |    37.5 |   55.55 | 22,29-31,40,45-50,55-58,64-69
 src/pages/dashboard/index.tsx                 |   68.75 |    28.57 |   42.85 |   67.74 | 21-23,30-36,54-87
  formatDate.ts             |     100 |    88.88 |     100 |     100 | 7*

### Packages Added:

1. **Framer Motion:** framer-motion
2. **Next UI:** @nextui-org/react"

### Demo:
https://github.com/Real-Dev-Squad/tiny-site-frontend/assets/70854507/0dfb95e6-73cd-47d0-9746-8a51f1d0afb2

https://github.com/Real-Dev-Squad/tiny-site-frontend/assets/70854507/84131400-bdb3-4af8-9fde-4520d26e0a69

### CODE REVIEW

File: __tests__/components/Toast.test.tsx [hunk 1]
```diff
@@ -7,20 +7,91 @@ describe('Toast', () => {
...
+                message="This is a toast message"
+                isVisible={true}
+                timeToShow={5000}
+                onDismiss={onDismiss}
+                type="success"
+            />
+        );
 
         expect(screen.getByText('This is a toast message')).toBeInTheDocument();
+        expect(screen.getByTestId('toast')).toBeInTheDocument();
```
Comment (heyrandhir) [inline __tests__/components/Toast.test.tsx:21]: I think this check should be done before checking the toast message. i.e the previous line should be after the current line.

File: __tests__/components/Toast.test.tsx [hunk 2]
```diff
@@ -7,20 +7,91 @@ describe('Toast', () => {
...
+                message="This is a toast message"
+                isVisible={true}
+                timeToShow={5000}
+                onDismiss={onDismiss}
+                type="error"
+            />
+        );
+
+        expect(screen.getByText('This is a toast message')).toBeInTheDocument();
+        expect(screen.getByTestId('toast')).toBeInTheDocument();
```
Comment (heyrandhir) [inline __tests__/components/Toast.test.tsx:64]: in the case of success as well we are checking these 2 lines only. anyway we can differentiate between success, info and failure toasts ?

File: __tests__/pages/dashboard.test.tsx [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+    });
+
+    it.skip('navigates to create page when "Create one" link is clicked', async () => {
+        render(<Dashboard />);
+        const createLink = screen.getByText('Create New');
+        fireEvent.click(createLink);
+
+        await waitFor(() => {
+            expect(screen.getByText('Enter a URL to shorten')).toBeInTheDocument();
+        });
```
Comment (heyrandhir) [inline __tests__/pages/dashboard.test.tsx:35]: why are both the tests skipped ?

File: __tests__/utils/formatDate.test.ts [hunk 4]
```diff
@@ -12,27 +12,57 @@ afterEach(() => {
...
+    it('should return Aug 5, 2021', () => {
+        const result = formatDate({
+            inputDate: '2021-08-05T00:00:00.000Z',
+            relativeDuration: false,
+            fullDate: false,
+        });
+        expect(result).toEqual('August 5, 2021');
+    });
+
+    // Add more test cases based on the function's behavior
```
Comment (heyrandhir) [inline __tests__/utils/formatDate.test.ts:59]: is this comment still relevant or can be removed ?

File: src/components/Dashboard/UrlListItem.tsx [hunk 5]
```diff
@@ -15,48 +16,52 @@ interface UrlListItemProps {
...
+                    >
+                        {TINY_SITE}/{url.ShortUrl}
+                    </Link>
+                    <Button
+                        className="rounded-full bg-gray-100 p-1.5 transition-all duration-75 hover:scale-105 hover:bg-blue-100 active:scale-95"
+                        onClick={() => copyButtonHandler(url.OriginalUrl)}
+                        testId="copy-button"
+                    >
+                        <span className="sr-only">Copy</span>
+                        <MdOutlineContentCopy style={{ color: 'black' }} />
```
Comment (heyrandhir) [inline src/components/Dashboard/UrlListItem.tsx:39]: can we use className in place of inline css

File: src/components/Toast/index.tsx [hunk 6]
```diff
@@ -1,33 +1,52 @@
...
+            animate={{ opacity: 1, y: 0 }}
+            exit={{ opacity: 0, y: '100%' }}
+            transition={{ duration: 0.5 }}
+            className={`fixed top-16 right-4 p-4 rounded-md shadow-md ${isVisible ? 'visible' : 'invisible'} ${
+                type === 'success'
+                    ? 'bg-green-500 text-white'
+                    : type === 'error'
+                    ? 'bg-red-500 text-white'
+                    : 'bg-blue-500 text-white'
+            }`}
```
Comment (heyrandhir) [inline src/components/Toast/index.tsx:30]: to many if else can we simplify this 
```js
const typeClassMap = {
  success: 'bg-green-500 text-white',
  error: 'bg-red-500 text-white',
  default: 'bg-blue-500 text-white', 
};

const typeClass = typeClassMap[type] || typeClassMap.default;
```

File: src/pages/app/index.tsx [hunk 7]
```diff
@@ -53,13 +72,14 @@ const App = () => {
 
     const handleUrl = () => {
         if (!isLoggedIn) {
-            displayErrorMessage('Not logged in');
+            setShowLoginModal(true);
+            if (url) localStorage.setItem('url', url);
```
Comment (heyrandhir) [inline src/pages/app/index.tsx:76]: why are we storing the url in local storage. can we not use a state variable for the same ?

**Review by heyrandhir — CHANGES_REQUESTED** [review]

File: src/pages/app/index.tsx (hunk 7 above)
Comment (sahsisunny) [reply src/pages/app/index.tsx:76]: No , after login the callback url redirect to default page.

File: __tests__/pages/dashboard.test.tsx (hunk 3 above)
Comment (sahsisunny) [reply __tests__/pages/dashboard.test.tsx:35]: Both tests are based on API calls, which is why I skipped them.

File: src/components/Toast/index.tsx (hunk 6 above)
Comment (sahsisunny) [reply src/components/Toast/index.tsx]: Done

File: src/components/Dashboard/UrlListItem.tsx (hunk 5 above)
Comment (sahsisunny) [reply src/components/Dashboard/UrlListItem.tsx]: Done

File: __tests__/utils/formatDate.test.ts (hunk 4 above)
Comment (sahsisunny) [reply __tests__/utils/formatDate.test.ts]: Removed

File: __tests__/components/Toast.test.tsx (hunk 2 above)
Comment (sahsisunny) [reply __tests__/components/Toast.test.tsx]: Done

File: __tests__/components/Toast.test.tsx (hunk 1 above)
Comment (sahsisunny) [reply __tests__/components/Toast.test.tsx:20]: Fixed

File: __tests__/pages/dashboard.test.tsx (hunk 3 above)
Comment (heyrandhir) [reply __tests__/pages/dashboard.test.tsx:35]: so in future we will remove the `skip` once API is ready ?

**Review by heyrandhir — APPROVED** [review]

File: __tests__/pages/dashboard.test.tsx (hunk 3 above)
Comment (sahsisunny) [reply __tests__/pages/dashboard.test.tsx:35]: Yes

**Review by vinit717 — APPROVED** [review]

**Review by heyrandhir — APPROVED** [review]

**Pull request merged**